Vision

TheFlameFish opened the pull request:
Adds vision to the robot with real and sim implementations.

Vision real was tested yesterday but has not been tested on the latest version.
Vision sim has been tested and appears to be functional with the modifications to moduleIOSim described in #9, although said modification has not been committed to this branch so sim will currently fail.

TheFlameFish committed:
Add Constants, move Vision constants to Constants, add AprilTagFieldLayout

TheFlameFish committed:
Installed PhotonVision 2025 beta 8

TheFlameFish committed:
Copied AprilTagVisionIO interface from 2024 codebase. Prone to modification.

TheFlameFish committed:
Made stub for AprilTagVisionIOReal

TheFlameFish committed:
Adapted Vision RealIO from 2024

TheFlameFish committed:
Rename SelfContainedPoseEstimator -> CameraPoseEstimator for clarity and because it's somewhat shorter.

TheFlameFish committed:
Add AdvantageKit dependencies to build.gradle so AutoLog annotations will work

TheFlameFish committed:
Update RobotContainer for renamed class CameraPoseEstimator

TheFlameFish committed:
Implement processing of vision data in AprilTagVision.java

TheFlameFish committed:
Made arrays in AprilTagVisionIOReal scale with the number of cameras

TheFlameFish committed:
Made Vision more self-contained (drive doesn't need to call a method from it) and minor doc changes

TheFlameFish committed:
Made AprilTagVision more self-contained (it now instantiates its own IO.)

TheFlameFish committed:
Bring vision up-to-date with main (Drive code). Not fully integrated with Drive yet.

TheFlameFish committed:
Have Drive instantiate Vision

TheFlameFish committed:
Minor formatting changes on Constants

TheFlameFish committed:
Added vision constants for Dory. I do not know the camera offsets yet.

TheFlameFish committed:
More minor formatting stuff in Constants.

TheFlameFish committed:
Dory camera constants

TheFlameFish committed:
Clean up formatting in AprilTagVision

TheFlameFish committed:
Add docs for VisionIOReal getEstimatedPoseUpdates method

TheFlameFish committed:
Merge from main -> vision, update WPILib to 2025.2.1

TheFlameFish committed:
Change vision to use the built-in AprilTagFieldLayout added in WPILib2025.2.1

TheFlameFish committed:
Get rid of unused vision variable in RobotContainer

TheFlameFish committed:
Add break statements to ends of switches for robotMode in AprilTagVision, made field bounds check a little more tolerant on the Z axis.

TheFlameFish committed:
Make AprilTagVisionIO work with 4 cameras (I hope to eventually make it adapt to any # of cameras), made default stdDevs (if the apriltag is not a known ID) very large so it'll be essentially discarded.

TheFlameFish committed:
Update PhotonLib to a stable version (2025.1.1)

TheFlameFish committed:
Merge from main to vision & apply spotless formatting

TheFlameFish committed:
Have CameraPoseEstimator directly take a PhotonCamera rather than instantiating the camera on its own.

TheFlameFish committed:
Move CameraPoseEstimator to AprilTagVisionIO so it can be used in sim

TheFlameFish committed:
Merge main into vision for remove BuildConstants from VCS

TheFlameFish committed:
Constants will switch between REAL and REPLAY depending on RobotBase.isReal()

TheFlameFish committed:
Add VisionSim (sim is not currently functional due to issues in ModuleIOSim but VisionSim has been tested with band-aid for ModuleIOSim)

TheFlameFish committed:
Merge main into vision (CI)

nolanhergert commented on src/main/java/frc/robot/subsystems/apriltagvision/AprilTagVisionIOReal.java:
```diff
@@ -0,0 +1,73 @@
+            poseArray[index] = estimatedRobotPose.estimatedPose;
+            timestampArray[index] = estimatedRobotPose.timestampSeconds;
+            Matrix<N3, N1> stdDevs =
+                getEstimationStdDevs(estimatedRobotPose, CameraResolution.HIGH_RES);
```
Assumes a HIGH_RES camera for all cameras, which is currently not the case on Dory. Can you update the CameraPoseEstimator class to include the resolution and index into it to determine resolution?

nolanhergert commented on src/main/java/frc/robot/subsystems/apriltagvision/AprilTagVisionIOReal.java:
```diff
@@ -0,0 +1,73 @@
+   * </br> If no updates are available, it will set defaults for `poseArray`, `timestampArray`, and
+   * `latencyArray`.
+   */
+  public void getEstimatedPoseUpdates() {
```
Is it possible to extract some/most of this into AprilTagVision or AprilTagVisionIO so that we don't have duplicate code between real and sim?

TheFlameFish replied:
I've moved the method to AprilTagVisionIO in [cb316e0](https://github.com/AEMBOT/FRC_2025/pull/10/commits/cb316e0bebb9e7b3316084cb1b926d5557901ac0), but I'm not sure I'm entirely satisfied with my solution. It passes the arrays that need to be updated into the method, the method updates them and does not return.

nolanhergert replied:
I would get rid of the boolean argument for doLatency as you are planning on doing latency measurement still for sim

nolanhergert commented on src/main/java/frc/robot/subsystems/apriltagvision/AprilTagVisionIO.java:
```diff
@@ -0,0 +1,157 @@
+            }) {
+      estStdDevs = VecBuilder.fill(Double.MAX_VALUE, Double.MAX_VALUE, Double.MAX_VALUE);
+    } else {
+      estStdDevs = estStdDevs.times(1 + (avgDist * avgDist / 20));
```
I don't understand the calculation here or above, both in single and multi-tag cases. It would be something we can add unit tests for too. However, in my opinion in the short term the potential imprecision is not as important as checking in mostly-working vision code to main that unblocks others.

A few options:
* Have a static std dev
* Check in what is here already (and used last year) and open an issue to revisit later.

I'm leaning towards making an issue. Thoughts?

TheFlameFish replied:
I agree that the second option would be better.

nolanhergert replied:
Created #11. Closing

nolanhergert commented on src/main/java/frc/robot/subsystems/apriltagvision/AprilTagVision.java:
```diff
@@ -0,0 +1,139 @@
+                  LightcycleCameras.rightCamToRobot,
+                  AprilTagConstants.poseStrategy)
+            };
+      case BUNNYBOT:
```
Would prefer to replace references to bunnybot with dory

TheFlameFish replied:
I agree but wasn't sure if that should be done in this PR given I think it affects all branches currently.

TheFlameFish committed:
Modified vision to account for differing camera resolutions

TheFlameFish committed:
Moved getEstimatedPoseUpdates() to AprilTagVisionIO instead of having it implemented un both Real and Sim.

TheFlameFish committed:
Remove boolean argument doLatency for getEstimatedPoseUpdates(), instead check robot mode in the method.

TheFlameFish commented:
Tested on real dory. Appears to be functioning well.

Ace5tar commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
looks like it is unused in drive, i dont see any reason to keep these additions unless you have another one

Ace5tar reviewed (approved):
looks good, I dont see any actual implementation outside of the subsysthem though, so getting it to integrate with odometry is still something that needs to be done

TheFlameFish commented:
> looks good, I dont see any actual implementation outside of the subsysthem though, so getting it to integrate with odometry is still something that needs to be done

It's instantiated in drive and adds vision measurements to the drivetrain's pose estimator if that's what you mean

TheFlameFish commented:
Judging by issues we're having while testing pathing, the camera offsets may be off.

Ace5tar committed:
changed camera offset from robot to be more accurate, still needs better measuring befor comp

Ace5tar commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -54,6 +55,8 @@ public class Drive extends SubsystemBase {
   private final SysIdRoutine moduleSteerRoutine;
   private final SysIdRoutine driveRoutine;
 
+  private AprilTagVision aprilTagVision = new AprilTagVision(poseEstimator);
```
not sure why this exists, seems to serve no purpose?

TheFlameFish replied:
I think it serves a purpose in that it instantiates vision. If you're referring to it being instantiated in drive instead of RobotContainer, that's mostly because it requires a drivetrain pose estimator which is instantiated by drive since it requires some suppliers from drive. 
   I think me and @nolanhergert had a conversation about instantiating it in RobotContainer vs in Drive and Nolan convinced me it would be better to do it in drive. However, neither of us can remember exactly why we made this decision. If preferred, it could be instantiated in RobotContainer (in which case we would instantiate Drive and then give AprilTagVision a reference to Drive's pose estimator).

Ace5tar reviewed (approved):
other than the seemingly unneccessary code in Drive.java, looks good

TheFlameFish committed:
Switched the axes on front camera offsets, changed camera offset variable names for clarity.

TheFlameFish committed:
Continuation of previous commit (change cam offset names)

TheFlameFish committed:
If camera doesn't see anything, output NaN instead of 0

TheFlameFish commented:
Tested on real robot. Appears to be working optimally.
Issue #11 is still an issue but we may want to check vision in without a solution and work on it later.

TheFlameFish commented:
@Ace5tar @nolanhergert 
I've made a few minor changes. Could you re-review it so it may be checked in (or further developed if deemed necessary)?

nolanhergert commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -88,4 +102,95 @@ public static final class Module {
+      NORMAL
+    }
+
+    public final class LightcycleCameras {
```
The name should be DoryCameras, right?

TheFlameFish replied:
Fixed in [7e8f234](https://github.com/AEMBOT/FRC_2025/pull/10/commits/7e8f2342714b4110aa87c5517b27379762b89663)

nolanhergert commented on src/main/java/frc/robot/subsystems/apriltagvision/AprilTagVision.java:
```diff
@@ -0,0 +1,147 @@
+  public AprilTagVision(SwerveDrivePoseEstimator swerveDrivePoseEstimator) {
+    CameraPoseEstimator[] visionPoseEstimators = {};
+    switch (currentRobot) {
+      case LIGHTCYCLE:
```
Change names to Dory and Nautilus

TheFlameFish replied:
Shouldn't PR #14 go through before I do that? If Vision is merged first, this could be done on PR #14.

nolanhergert replied:
Agree, #14 should go through first. This is more in preparation for that point.

nolanhergert commented on src/main/java/frc/robot/subsystems/apriltagvision/AprilTagVision.java:
```diff
@@ -0,0 +1,147 @@
+                  new PhotonCamera(LightcycleCameras.frontCamName),
+                  LightcycleCameras.frontCamToRobot,
+                  AprilTagConstants.poseStrategy,
+                  CameraResolution.HIGH_RES),
```
4 high res on nautilus and 4 low res on dory

TheFlameFish replied:
Fixed in [7e8f234](https://github.com/AEMBOT/FRC_2025/pull/10/commits/7e8f2342714b4110aa87c5517b27379762b89663)

nolanhergert commented on src/main/java/frc/robot/subsystems/apriltagvision/AprilTagVisionIO.java:
```diff
@@ -0,0 +1,218 @@
+            }
+            latencyArray[index] = Timer.getFPGATimestamp() - camResult.get().getTimestampSeconds();
+          },
+          () -> {
```
Maybe add a comment: 
`// Set values to defaults if no apriltags found`

TheFlameFish replied:
Fixed in [2005e84](https://github.com/AEMBOT/FRC_2025/pull/10/commits/2005e846b46aa9bce27d2d607271e5b3a60b6ca1)

TheFlameFish committed:
Minor documentation

TheFlameFish committed:
Apply spotless

TheFlameFish committed:
Add (kinda placeholder) Nautilus constants

TheFlameFish committed:
Refactor internal robot names

TheFlameFish committed:
Resolve main conflict

TheFlameFish commented:
Tested on Dory and functional. Issues #11 and #16 are concerning but we may want to merge vision to main now to unblock other development, working on the issues later.

Ace5tar reviewed (approved):
i dont see any issues especially if it has been tested and works good 👍

TheFlameFish committed:
Log stdDevs as NaN when no targets vision

TheFlameFish committed:
Adjusted Dory camera offsets

GreenTomato5 commented:
Hey just letting y'all know there seems to be a bug in your sim code (we really like how y'all structure april tag sim so we were implementing ours the way y'all do 😅 ). You dont initialize timestamp array in the sim io layer and then don't update it in sim (bc latency is cooked in sim) so when you reference it you get a null reference error and code crashes

TheFlameFish committed:
Set camera offsets for Nautilus

TheFlameFish committed:
Merge remote-tracking branch 'origin/main' into vision

TheFlameFish committed:
Get vision working on Nautilus

nolanhergert commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -191,23 +200,23 @@ public final class DoryCameras {
               new Translation3d(
-                  Units.inchesToMeters(11), Units.inchesToMeters(7), Units.inchesToMeters(4)),
-              new Rotation3d(Units.degreesToRadians(180), Units.degreesToRadians(-30), 0.0));
+                  Units.inchesToMeters(11), Units.inchesToMeters(-7), Units.inchesToMeters(4)),
```
This negation on y is still odd to me that it's only on Dory front cameras and not on the back cameras or Nautilus.

I would leave a comment indicating that this is unexpected behavior so it's easy to notice in the future.

TheFlameFish replied:
Documentation added: 4d1f94fc2b2b816a8f3df132e59f8eae8976f183

TheFlameFish committed:
Documentation changes
   Remove outdated TODO
   Add TODO, figure out Dory front camera position Y inversion

TheFlameFish committed:
Have vision sim set latency to 0 instead of null

TheFlameFish committed:
Have drive call updatePose for vision sim

TheFlameFish committed:
Switch sim to use Nautilus constants

TheFlameFish committed:
Set drive posestimator position in sim

TheFlameFish commented:
Note that we should have #26 merge before merging this; it has new `AprilTagFieldLayout`s for the welded and AndyMark fields

TheFlameFish commented:
> Hey just letting y'all know there seems to be a bug in your sim code (we really like how y'all structure april tag sim so we were implementing ours the way y'all do 😅 ). You dont initialize timestamp array in the sim io layer and then don't update it in sim (bc latency is cooked in sim) so when you reference it you get a null reference error and code crashes

Thank you for finding that! Did 4baf52e3e38a9a45dc9f53b373093db3cb6a9aeb resolve it?

TheFlameFish committed:
Merge branch 'main' into vision

TheFlameFish commented:
I believe this is ready for merge into main.

TheFlameFish commented:
Tested on Nautilus and working optimally.

@Ace5tar could you do a final review?
